examples/nlp/semantic_similarity_with_bert.py

```diff
@@ -0,0 +1,362 @@
+
+        # set to true if data generator is used for training/validation
+        if self.include_targets:
+            labels = np.array(self.labels[indexes], dtype="int32")
+            return [input_ids, masks, token_type_ids], labels
+        else:
+            return [input_ids, masks, token_type_ids]
+
+    def on_epoch_end(self):
+        # Shuffle indexes after each epoch if shuffle is set to True.
+        self.indexes = np.arange(len(self.sentence1))
+        if self.shuffle:
+            np.random.RandomState(42).shuffle(self.indexes)
+
+
+"""
+## Build The Model
+"""
+
+
+def build_model():
+    # Encoded token ids from BERT tokenizer.
+    input_ids = tf.keras.layers.Input(
+        shape=(max_length,), dtype=tf.int32, name="input_ids"
+    )
+    # Attention masks indicates to the model which tokens should be attended to.
+    attention_masks = tf.keras.layers.Input(
+        shape=(max_length,), dtype=tf.int32, name="att_mask"
+    )
+    # Token type ids are binary masks identifying different sequences in the model.
+    token_type_ids = tf.keras.layers.Input(
+        shape=(max_length,), dtype=tf.int32, name="token_type_ids"
+    )
+    # Loading pretrained BERT model.
+    bert_model = transformers.TFBertModel.from_pretrained("bert-base-uncased")
+    sequence_output, pooled_output = bert_model(
+        input_ids, attention_mask=attention_masks, token_type_ids=token_type_ids
+    )
+    # Applying hybrid pooling approach to bert sequence output.
+    avg_pool = tf.keras.layers.GlobalAveragePooling1D()(sequence_output)
+    max_pool = tf.keras.layers.GlobalMaxPooling1D()(sequence_output)
+    concat = tf.keras.layers.concatenate([avg_pool, max_pool])
+    dropout = tf.keras.layers.Dropout(0.3)(concat)
+    output = tf.keras.layers.Dense(3, activation="softmax")(dropout)
+
+    model = tf.keras.models.Model(
+        inputs=[input_ids, attention_masks, token_type_ids], outputs=output
+    )
+
+    model.compile(
+        optimizer=tf.keras.optimizers.Adam(learning_rate=learning_rate),
+        loss="categorical_crossentropy",
+        metrics=["acc"],
+    )
+    return model
+
+
+"""
+Create the model under a distribution strategy scope.
+"""
+# Build model with distributed strategy.
+strategy = tf.distribute.MirroredStrategy()
+with strategy.scope():
+    model = build_model()
+
+print(f"Strategy: {strategy}")
+model.summary()
+
+"""
+Create train and validation data generators
+"""
+train_data = BertSemanticDataGenerator(
+    train_df.sentence1.astype("str"),
+    train_df.sentence2.astype("str"),
+    y_train,
+    batch_size=batch_size,
+    shuffle=True,
+)
+valid_data = BertSemanticDataGenerator(
+    valid_df.sentence1.astype("str"),
+    valid_df.sentence2.astype("str"),
+    y_val,
+    batch_size=batch_size,
+    shuffle=False,
+)
+
+"""
+## Train the Model
+"""
+history = model.fit_generator(train_data, validation_data=valid_data, epochs=epochs,)
```

Comment:
A big problem here is that you're not doing fine-tuning properly.

To do fine-tuning, you need 2 steps:

1. Freeze `bert_model` and fit the model. This is called "feature extraction": you just reuse the pretrained features without modifying them.
2. unfreeze `bert_model` and fit again (with a very small learning rate). This is called fine-tuning.

Step 2 may not be necessary -- for many applications, feature extraction is sufficient.

Also, please note that:

- You can pass a generator to `fit`. `fit_generator` is legacy.
- If you fit from a generator, you should generally use data multiprocessing to get better GPU utilization. See the docs for `fit()`.

Reply:
I have used only "feature extraction" and trained by freezing `bert_model` that gave me accuracy around 37%.
After some tuning, I experimented stacking a classification head below freezed bert_model :
1. stacking 2 dense layers with dropout gave accuracy around 45%.
2. stacking 1 Bi-LSTM layer with dropout gave accuracy around 69%.

`Accuracy of unfreeze bert_model (our current model) is around 87%.`

I have gone through this [paper](https://arxiv.org/pdf/1911.03090.pdf) yesterday as they have experimented with different datasets and models, I found this result related to our task.

```
Frozen up to  MNLI     MNLI-mm
0th           84.2      82.0
9th           82.0      82.4
12th          56.4      57.1
```
--------
I have changed fit_generator to fit and also used multiprocessing for better GPU utilization.

Reply:
It makes sense that you get better results by training the entire BERT model, but in this case it makes little sense to load a pretrained model (since you lose its representations almost immediately due to the presence of untrained layers in the model).

Either load the BERT model from scratch (no pretrained weights), or do actual fine-tuning (the two steps I described above). The first option is the simplest.

Reply:
I got your point for the first time but I was little bit concerned about the accuracy. Now, the model is trained on both the steps  "feature extraction" and "fine-tuning" (unfreeze bert_model)  that gave quite good results.
I have also changed the data generator and used batch_encode_plus that has speed up the training.